pennylane/gradients/gradient_transform.py

```diff
@@ -394,80 +397,92 @@ def reorder_grads(grads, tape_specs):
-        def _reshape(x):
-            return qml.math.reshape(x, (1,) if x.shape == () else (1, -1))
-
-        if not (multi_meas or has_partitioned_shots):
-            # Single parameter, single measurements, no shot vector
-            return tdot(_reshape(qjac), cjac)
-
-        if not (multi_meas and has_partitioned_shots):
-            # Single parameter, multiple measurements or shot vector, but not both
-            return tuple(tdot(_reshape(q), cjac) for q in qjac)
-
-        # Single parameter, multiple measurements, and shot vector
-        return tuple(tuple(tdot(_reshape(_q), cjac) for _q in q) for q in qjac)
-
-    if not multi_meas:
-        # Multiple parameters, single measurement
-        qjac = stack(qjac)
-        if not cjac_is_tuple:
-            cjac = stack(cjac)
-            if has_partitioned_shots:
-                return tuple(tdot(stack(q), cjac) for q in qjac)
-            return tdot(qjac, cjac)
-        if has_partitioned_shots:
-            return tuple(tuple(tdot(q, c) for c in cjac if c is not None) for q in qjac)
-        return tuple(tdot(qjac, c) for c in cjac if c is not None)
-
-    # Multiple parameters, multiple measurements
-    if not cjac_is_tuple:
-        cjac = stack(cjac)
-        if has_partitioned_shots:
-            return tuple(tuple(tdot(stack(_q), cjac) for _q in q) for q in qjac)
-        return tuple(tdot(stack(q), cjac) for q in qjac)
-    if has_partitioned_shots:
-        return tuple(
-            tuple(tuple(tdot(stack(_q), c) for c in cjac if c is not None) for _q in q)
-            for q in qjac
-        )
-    return tuple(tuple(tdot(stack(q), c) for c in cjac if c is not None) for q in qjac)
+    if tape.shots.has_partitioned_shots:
+        return tuple(_single_shots_contraction(qjac_s, cjac, tape) for qjac_s in qjac)
+
+    return _single_shots_contraction(qjac, cjac, tape)
+
+
+def _single_shots_contraction(qjac, cjac, tape):
+    single_trainable_param = len(tape.trainable_params) == 1
+    if len(tape.measurements) == 1:
+        return _single_meas_contraction(qjac, cjac, single_trainable_param)
+
+    return tuple(_single_meas_contraction(qj, cjac, single_trainable_param) for qj in qjac)
+
+
+def _single_meas_contraction(qjac, cjac, single_trainable_param):
+    if single_trainable_param:
+        # return spec squeezed out singleton dimension, so we need to add it back in
+        qjac = qml.math.expand_dims(qjac, axis=0)
+    qjac = qml.math.stack(qjac)
+
+    # pytrees are easiest way to handle different argnum conventions: argnum=None, argnum=0, argnum=[0], etc.
+    cjac_leaves, struct = flatten(cjac)
+
+    contracted = []
+    for cjac_for_arg in cjac_leaves:
+        tdot = qml.math.tensordot(cjac_for_arg, qjac, axes=[[0], [0]])
+        # not sure how to rationalize this tranposition, but its needed
+        contracted.append(tdot if len(tdot.shape) == 0 else qml.math.moveaxis(tdot, -1, 0))

```

Comment:
I'm happy that @PietropaoloFrisoni  confirmed that this eventually fixes the issue. But I still feel what happened to the contraction to each leaf is so confusing: let's say for some leaf we have `cjac_for_arg` of shape `(a, b, c)` and `qjac` is of shape `(d, e, f, g)`, then what we gonna do is first we tdot them to get a new tensor of `(b, c, e, f, g)` and cyclic rotate axes to reach the final shape `(g, b, c, e, f)`?
I'll approve any way since bug fixed but just got the feeling that we might need to revisit some day eventually...

Reply:
What are the common cases for the shapes of `cjac_for_arg` and `qjac`? Would it be possible that actually they are either 1 axis dotting 1 axis or 2 axes dotting 2 axes so that for the latter case we were just doing a usual transposition

Reply:
I think its something like `(trainable_param_idx, *qnode_arg_shape)` and `(trainable_param_idx, *measurement_shape)`? Then we sum over `trainable_param_idx`. But I guess this does not explain that transposition at all.

Reply:
Yeah, I just needed to contract in the opposite order.